Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -240,9 +240,16 @@ void GrpcStream::FastFinishCompletionsBlocking() {
     completion->Cancel();
   }
 
+  bool queue_shutting_down = worker_queue_->is_shutting_down();
+
   for (auto completion : completions_) {
     // This is blocking.
     completion->WaitUntilOffQueue();
+
+    // If the queue is shutting down it won't allow any completions to fire.
+    if (queue_shutting_down) {
```

Comment:
Honestly, while this change is certainly the right thing to do, the fact that ownership is kinda vague here seems a little concerning. Do you think this is something we could improve long-term? Perhaps add some sort of a TODO?

Reply:
I added a TODO.

The refactoring that's required was more than I wanted to attempt here. The outline of a solution is that since we need to `new`/`delete` for the purposes of sharing with gRPC, `GrpcCompletion` objects should only exist for that purpose and should not otherwise be shared.

Concretely, this means that when the completion is called by gRPC, it should package up the values it wants to pass into the closure it submits onto the AsyncQueue rather than passing a pointer to itself. This would allow it to  unconditionally, immediately delete itself and clear this up.

I didn't make this change because (among other things) I don't entirely understand the relationship between the lifetime of the message and the callback onto our queue. Let's discuss and address in a separate PR.

Reply:
SGTM.

Reply:
The ASAN test failures indicate that while the old mode leaked on shutdown, the new mode uses-after-free. Neither are good.

I've reworked this to use `shared_ptr`. This allows either gRPC's hold or `GrpcStream`'s hold to exceed the other without negative consequence. See comments for exactly how this works.